assets/js/googlesitekit/data/create-fetch-store.test.js

```diff
@@ -0,0 +1,328 @@
+				reducerCallback: ( state ) => state,
+			} );
+
+			expect( Object.keys( fetchStoreDefinition.actions ) ).toEqual( [
+				'fetchSaveSomeData',
+				'receiveSaveSomeData',
+			] );
+		} );
+
+		describe( 'fetch', () => {
+			it( 'validates parameters based on argsToParams', () => {
+				const consoleErrorSpy = jest.spyOn( global.console, 'error' );
+
+				dispatch.fetchGetSomeData();
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'objParam is required.' );
+
+				dispatch.fetchGetSomeData( 123 );
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'objParam is required.' );
+
+				dispatch.fetchGetSomeData( {} );
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'aParam is required.' );
+
+				consoleErrorSpy.mockClear();
+			} );
+
+			it( 'yields the expected actions for an arguments error', () => {
+				const fetchStoreDefinition = createFetchStore( {
+					baseName: 'SaveSomeData',
+					controlCallback: async () => true,
+					reducerCallback: ( state ) => state,
+					argsToParams: ( requiredParam ) => {
+						invariant( requiredParam, 'requiredParam is required.' );
+						return {
+							requiredParam,
+						};
+					},
+				} );
+
+				const action = fetchStoreDefinition.actions.fetchSaveSomeData();
+
+				// Catch invariant to get exactly the error we expect.
+				let error;
+				try {
+					invariant( false, 'requiredParam is required.' );
+				} catch ( err ) {
+					error = err;
+				}
+
+				expect( action.next().value ).toEqual( {
+					response: undefined,
+					error,
+				} );
+			} );
+
+			it( 'yields the expected actions for a success request', () => {
+				const fetchStoreDefinition = createFetchStore( {
+					baseName: 'SaveSomeData',
+					controlCallback: async () => true,
+					reducerCallback: ( state ) => state,
+				} );
+
+				const action = fetchStoreDefinition.actions.fetchSaveSomeData();
+
+				expect( action.next().value.type ).toEqual( 'START_FETCH_SAVE_SOME_DATA' );
+				expect( action.next().value.type ).toEqual( 'FETCH_SAVE_SOME_DATA' );
+				expect( action.next( 42 ).value.type ).toEqual( 'RECEIVE_SAVE_SOME_DATA' );
+				expect( action.next().value.type ).toEqual( 'FINISH_FETCH_SAVE_SOME_DATA' );
+				expect( action.next().value ).toEqual( {
+					response: 42,
+					error: undefined,
+				} );
+			} );
+
+			it( 'yields the expected actions for an error request', () => {
+				const fetchStoreDefinition = createFetchStore( {
+					baseName: 'SaveSomeData',
+					controlCallback: async () => true,
+					reducerCallback: ( state ) => state,
+				} );
+
+				const action = fetchStoreDefinition.actions.fetchSaveSomeData();
+
+				const error = {
+					code: 'this-went-wrong',
+					message: 'This went wrong.',
+				};
+
+				expect( action.next().value.type ).toEqual( 'START_FETCH_SAVE_SOME_DATA' );
+				expect( action.next().value.type ).toEqual( 'FETCH_SAVE_SOME_DATA' );
+				expect( action.throw( error ).value.type ).toEqual( 'CATCH_FETCH_SAVE_SOME_DATA' );
```

Comment:
Same as above.

Reply:
See above, how would we trigger an error like this by just destructuring the generator?

Reply:
The error would be triggered when invoking the generator function - that part would still happen. It's the response that you can destructure into a list of all of its `yield`ed values as an array which can be listed out into variables that you can make the same assertions on, but in a more readable way. I'll take a closer look and give you a concrete example in a bit, unless there's something I'm missing here that would prevent this from being possible.

Reply:
You're right, I misunderstood what was happening here 😄 

See my comment here https://github.com/google/site-kit-wp/pull/1388#discussion_r436688633

Reply:
This is okay for now, but going forward if we want to test a lot of async actions like this we might want to hook up a until like `redux-saga-tester` or similar to allow us to write more flexible generator tests. These tests are all fine but the issue and worth keeping, but if a _new_ yielded action is added in-between two actions here the tests will fail even though they're largely still behaving as-intended.

Having to add extra `action.next()` here when we modify the actions seems onerous to me. But that's only if we start writing a lot of tests around these behaviours. If these lower-level libraries have these types of lower-level tests, maybe that's okay. 🙂

Reply:
Yeah that's what I was thinking, we don't need to go that deep for most of our functionality, but since this is used in so many places, I wanted to make sure to have very solid coverage.